pkg/gameservers/missing.go

```diff
@@ -0,0 +1,157 @@
+func NewMissingPodController(health healthcheck.Handler,
+	kubeClient kubernetes.Interface,
+	agonesClient versioned.Interface,
+	kubeInformerFactory informers.SharedInformerFactory,
+	agonesInformerFactory externalversions.SharedInformerFactory) *MissingPodController {
+
+	podInformer := kubeInformerFactory.Core().V1().Pods().Informer()
+	gameServers := agonesInformerFactory.Agones().V1().GameServers()
+	c := &MissingPodController{
+		podSynced:        podInformer.HasSynced,
+		podLister:        kubeInformerFactory.Core().V1().Pods().Lister(),
+		gameServerSynced: gameServers.Informer().HasSynced,
+		gameServerGetter: agonesClient.AgonesV1(),
+		gameServerLister: gameServers.Lister(),
+	}
+
+	c.baseLogger = runtime.NewLoggerWithType(c)
+	c.workerqueue = workerqueue.NewWorkerQueue(c.syncGameServer, c.baseLogger, logfields.GameServerKey, agones.GroupName+".MissingPodController")
+	health.AddLivenessCheck("gameserver-missing-pod-workerqueue", healthcheck.Check(c.workerqueue.Healthy))
+
+	eventBroadcaster := record.NewBroadcaster()
+	eventBroadcaster.StartLogging(c.baseLogger.Debugf)
+	eventBroadcaster.StartRecordingToSink(&typedcorev1.EventSinkImpl{Interface: kubeClient.CoreV1().Events("")})
+	c.recorder = eventBroadcaster.NewRecorder(scheme.Scheme, corev1.EventSource{Component: "missing-pod-controller"})
+
+	gameServers.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		UpdateFunc: func(_, newObj interface{}) {
+			gs := newObj.(*agonesv1.GameServer)
+			_, isDev := gs.GetDevAddress()
+			if !isDev && !isBeforePodCreated(gs) && !gs.IsBeingDeleted() && !gs.IsUnhealthy() {
+				c.workerqueue.Enqueue(gs)
+			}
+		},
+	})
+
+	return c
+}
+
+// Run processes the rate limited queue.
+// Will block until stop is closed
+func (c *MissingPodController) Run(stop <-chan struct{}) error {
+	c.baseLogger.Debug("Wait for cache sync")
+	if !cache.WaitForCacheSync(stop, c.gameServerSynced, c.podSynced) {
+		return errors.New("failed to wait for caches to sync")
+	}
+
+	c.workerqueue.Run(1, stop)
+	return nil
+}
+
+func (c *MissingPodController) loggerForGameServerKey(key string) *logrus.Entry {
+	return logfields.AugmentLogEntry(c.baseLogger, logfields.GameServerKey, key)
+}
+
+// syncGameServer checks if a GameServer has a backing Pod, and if not,
+// moves it to Unhealthy
+func (c *MissingPodController) syncGameServer(key string) error {
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		// don't return an error, as we don't want this retried
+		runtime.HandleError(c.loggerForGameServerKey(key), errors.Wrapf(err, "invalid resource key"))
+		return nil
+	}
+
+	// check if the pod exists
+	if pod, err := c.podLister.Pods(namespace).Get(name); err != nil {
+		if !k8serrors.IsNotFound(err) {
+			return errors.Wrapf(err, "error retrieving Pod %s from namespace %s", name, namespace)
+		}
+		c.loggerForGameServerKey(key).Debug("Pod is missing. Moving GameServer to Unhealthy.")
```

Comment:
It seems like this log line should be after the if statement. If the pod is found and it's not a game server pod, then we won't get this log line but we will still try to move the game server to unhealthy.

Reply:
Good catch!